artifact/v3/manifest/manifest.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <artifact/v3/manifest/manifest.hpp>
+
+#include <string>
+#include <sstream>
+
+#include <common/error.hpp>
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace manifest {
+
+namespace io = mender::common::io;
+namespace error = mender::common::error;
+
+const size_t expected_shasum_length {64};
+const size_t expected_whitespace {2};
+
+ExpectedManifest Parse(mender::common::io::Reader &reader) {
+	Manifest m {};
+
+	stringstream ss {};
+
+	io::StreamWriter sw {ss};
+
+	auto err = io::Copy(sw, reader);
+	if (err) {
+		return expected::unexpected(err);
+	}
+
+	string entry_name {};
+	string shasum {};
+	string line {};
+
+	while (getline(ss, line, '\n')) {
+		/* Since the shasums are 64 character, and two whitespaces are required, and the shortest
+		 * name is 1 character */
+		if (line.size() < expected_shasum_length + expected_whitespace + 1) {
+			return expected::unexpected(error::MakeError(
+				error::GenericError, "Malformed input line. Line too short. Line: " + line));
```

Comment:
I agree with Kristian we should at least have some generic `ArtifactParseError` instead of using `GenericError` everywhere.